### ISSUE TITLE
Rewrite controller, database, registry, store bin/boot scripts in Go

### ISSUE THREAD

**Issue by carmstrong** [deis/deis#2701]
@aledbf Did a great job rewriting some components in Go, but we should move all components to a Go-based /bin/boot script. @gabrtv also prefers this method to calling confd, as we can vendor the code.

**Comment by bacongobbler** [comment]
I think I've mentioned this a few times casually in different channels, but I don't think this is
the way we need to go forward with the boot scripts. AFAIK there were three main requirements
coming into this proposal:

1) Make the codebase less fragile
2) Make the codebase easier to maintain
3) Add test coverage to gain confidence in changes

point 1 is hard to quantify without point 3. boot scripts are semi-impossible to test unless you
want to mock etcd endpoints, somehow monitor the filesystem inside the container for file changes,
monitor that the database is being backed up, etc. This will be the same if the boot scripts are in
bash or it's in Go. There is no win here on either side.

As for point 2, using something like go-extpoints actually makes the code *extremely* convoluted.
Hacking on different points in the init script in #3412 was **hard**, which left me running back to
shell faster than you can say "Go". Using extension points to share standard interfaces for a boot
script felt wrong. Sharing a library of standard functions used across multiple drivers == good.

However, I do believe that there is some things we can take from this that will hit on all three
points, and that is through separate refactoring:

1) the publish loops should not exist in bin/boot. They should be refactored into deis/publisher.
We have test infrastructure there for publishing apps to etcd, and we can do the same for the platform
components.
2) Long-running commands inside the boot script (such as "waiting for confd to write initial
templates") should be offloaded or be executed as a StartPre or a StartPost operation, which will
touch on #2906. The only thing bin/boot should be doing is firing up confd and the main process
itself. We already do some of this in [deis-builder](https://github.com/deis/deis/blob/b06b85a474262c643c61377bd756e768420c489a/deisctl/units/deis-builder.service#L12-L13) but obviously we can do better.

I'm always open for discussion on this and my mind can be swayed if there was a strong reason to
move in that direction, but I really feel that this a **bad** move. There are better ways we can
refactor these init scripts, and that starts with better separation of concerns.

TL;DR I'm :-1: on refactoring init scripts in Go.

**Comment by carmstrong** [comment]
We're going to schedule an architectural meeting to discuss this in-depth with @bacongobbler and @gabrtv. @aledbf you're more than welcome to attend!

**Comment by carmstrong** [comment]
Replaced by #3758
